Add decision for API design (#35)
vshn/application-catalog-docs: docs/modules/ROOT/pages/explanations/decisions/api-design.adoc

= Decision on API Design

== Problem

> This decision is about how we can achieve a "common look" as defined https://kb.vshn.ch/app-catalog/explanations/why_exists.html#_standard_way_to_provide_services[here].

Reviewer: ```suggestion
This decision is about how we can achieve a common look as defined https://kb.vshn.ch/app-catalog/explanations/why_exists.html#_standard_way_to_provide_services[here].
```

The ideal way would be, that we can re-use the exact same XRD for each set of services.
For example all implementations for PostgreSQL have exactly the same API for the user.
Due to the nature of the product, not all implementations provide the same functionality.
The set of available features is most likely not the same between most providers.

> Also with Crossplane we're a bit limited on what's possible with translating the source API spec into the target API spec.

Reviewer: That warrants a more detailed reasoning, it doesn't say too much this way...

== Requirements

> * APIs for a set of similar services should feel similar
> * They should only need minimal configuration for a default case
> * We should be able to leverage unique features of the service providers

Reviewer: ```suggestion
* APIs for a set of similar services feel similar
* Only minimal configuration is needed for a default case
* Unique features of a service provider can be leveraged
```

== Proposals

=== Proposal 1: Use same XRD with provider specific fields

There's exactly one XRD for a set of services.
This XRD contains a minimal set of common parameters, as well as a field for provider specific configurations.

.Example Claim
[source, yaml]
----
apiVersion: appcat.vshn.io/v1
kind: PostgreSQL
metadata:
  name: my-instance
spec:
  parameters:
    name: my-instance
    majorVersion: '14'
    pgSettings: {}
    exoscale:
      plan: hobbyist-2
      type: pg
      maintenance:
        day: 2
        hour: 22
        minute: 0
      backup:
        hour: 20
        minute: 0
    vshn:
      minorVersion: '4'
----

Advantages::

* We can use the same XRD
* Adding new providers to the list is easy

Disadvantages::

* We don't really have a "common look"
* The customer needs to know what the underlying service provider is, even for simple instances

> === Proposal 2: Separate XRDs

Reviewer: Maybe use the same semantics as Crossplane does? At the top-level (`spec.X`) _always_ have the same parameters, and then have a provider specific part like `spec.forProvider.X`.

Author: I'm not quite sure what you mean here. Both examples have the same parameters under `spec.parameters.*`?

Reviewer: My concern is, that `.spec.size` is has different options, depending on the provider used, which might be confusing. E.g. on one we have `.spec.size.plan` and on another we have `.spec.size.cpu`. So my proposal (maybe a new one) would be:

```
apiVersion: exoscale.appcat.vshn.io/v1
kind: PostgreSQL
metadata:
  name: my-instance
spec:
  parameters:
    name: my-instance
    version: '14'
    maintenance:
      day: 2
      hour: 22
      minute: 0
    backup:
      hour: 20
      minute: 0
    providerSpecific:
      service:
        type: pg
        settings: {}
        plan: hobbyist-2
```

This introduces the `providerSpecific` field, which is different for each provider, while the other are always the same.

Author: Added it as another proposal.

= Decision on API Design

== Problem

This decision is about how we can achieve a common look as defined https://kb.vshn.ch/app-catalog/explanations/why_exists.html#_standard_way_to_provide_services[here].

The ideal way would be, that we can re-use the exact same XRD for each set of services.
For example all implementations for PostgreSQL have exactly the same API for the user.
Due to the nature of the product, not all implementations provide the same functionality.
The set of available features is most likely not the same between most providers.
Also with Crossplane we're a bit limited on what's possible with translating the source API spec into the target API spec:

* No conditionals in compositions
* Only a limited set of transformations
* Can't create random values

[NOTE]
This is not a complete list of limitations.

== Requirements

* APIs for a set of similar services feel similar
* Only minimal configuration is needed for a default case
* Unique features of a service provider can be leveraged

== Proposals

=== Proposal 1: Use same XRD with provider specific fields

> 

Reviewer 2: I think over time this looks more like an anti-pattern of Kubernetes.
Consider adding AWS and GCP, eventually you'll end up with maximum 3 common parameters, the rest is very provider-specific and in `.spec.parameters.{aws,gcp,exoscale,vshn}` keys. Might as well create different CRDs for each provider and accept that those 3 common parameters are duplicated.

Author: I've added the remark as a disadvantage.

There's exactly one XRD for a set of services.
This XRD contains a minimal set of common parameters, as well as a field for provider specific configurations.

.Example Claim
[source, yaml]
----
apiVersion: appcat.vshn.io/v1
kind: PostgreSQL
metadata:
  name: my-instance
spec:
  parameters:
    name: my-instance
    majorVersion: '14'
    pgSettings: {}
    exoscale:
      plan: hobbyist-2
      type: pg
      maintenance:
        day: 2
        hour: 22
        minute: 0
      backup:
        hour: 20
        minute: 0
    vshn:
      minorVersion: '4'
----

Advantages::

* We can use the same XRD
* Adding new providers to the list is easy

Disadvantages::

* We don't really have a "common look"
* The customer needs to know what the underlying service provider is, even for simple instances

=== Proposal 2: Separate XRDs

This approach will have different XRDs for each provider.
However, they will follow the same structure, as far as possible.

> All the toplevel fields will be the same, the differences will be apparent in the sub-fields.

Reviewer 2: I would remove this artificial limitation, it doesn't make sense in provider-specific CRD.

Author: But this is the point of the second proposal.

If you look at the example you see that all parameters concerning the size of the instance are under the key `size` then under this key we use the provider specific naming, like plan, or CPU/Memory/Disk. Depending on what the provider actually understands.

Then you know that all configuration about the size for an instance is under the field `size`, thus giving them a common look. Because it may not be apparent that a `plan` or `profile` or whatever name the provider thinks of, is about sizing.

Reviewer 2: I understood the examples. But if a provider doesn't offer any customization about maintenance for example, then the whole `maintenance` top level key doesn't apply. Or if a service uses some sort of autoscaling and `size` wouldn't matter too.

Author: In that case we simply omit it. Then it's also clear from the API that this isn't supported (yet?) for the given provider.

Reviewer 2: fine by me to leave it as is.

They will also have the same kind, but a different APIVersion.

.Example Claim
[source, yaml]
----
# Exoscale example
apiVersion: exoscale.appcat.vshn.io/v1
kind: PostgreSQL
metadata:
  name: my-instance
spec:
  parameters:
    name: my-instance
    service:
      version: '14'
      type: pg
      settings: {}
    size:
      plan: hobbyist-2
    maintenance:
      day: 2
      hour: 22
      minute: 0
    backup:
      hour: 20
      minute: 0

# Converged example
apiVersion: vshn.appcat.vshn.io/v1
kind: PostgreSQL
metadata:
  name: my-instance
spec:
  parameters:
    name: my-instance
    service:
      version: '14.4'
      settings: {}
      pgbouncer:
        settings: {}
    size:
      cpu: 2
      memory: 8Gi
      disk: 200Gi
      encrypted: true
    maintenance:
      day: 2
      hour: 22
      minute: 0
    backup:
      hour: 20
      minute: 0
----

Advantages::

* Same look between all providers, deploying PostgreSQL to any backend feels familiar

> * Specifics can be handled in a sensible way

Reviewer 2: I wouldn't focus on "same look" too much. The same look can only apply if various providers share a common set of parameters. But then you'd have to agree on the same terminology. But depending on provider, some term might deviate from the DSL for the provider, maybe even mean the opposite.
As soon as one provider doesn't implement the "common" parameters then this "common look" is already outdated and sets expectations for the user only to be disappointed.

Author: See above comment. Maybe it's not quite clear what I mean with "common set of parameters". Think of them more of sets of more specific parameters.

For example most providers allow you to configure the service to some degree (version or settings). My proposal now states that all those params concerning the configuration of the service should go under `service`. Then, as a user, you know what these params are for. Same thing for the sizing or backups, etc. that might be common for a given service.

If a provider doesn't implement a given set of params at all, then we omit it.

Reviewer 2: I have no strong feelings about grouping some provider-specific settings into top-level parameter groups. But I would avoid having those groups present in the API if they don't apply. I would also avoid renaming fields that are not matching the provider's naming and terminology.
If you think having some same-named group across different DB services suffices as having a "common look", then that's fine by me.

Disadvantages::

* The user could get confused by objects with the same kind

=== Proposal 3: Composed XRDs

> 

Reviewer 2: tbh I would remove this proposal as its just not a real option.
Having a base CRDs just to generalize 3 parameters is, IMO, unacceptable for both the user and maintainers.

Author: I'd like to keep it in the page, just for completeness' sake.

There's a base XRD that handles the common configurations.
All provider specific parameters are put in a separate XRD.

This idea may not be possible with Crossplane, but it's here for completeness’ sake.

.Example Claim
[source, yaml]
----
apiVersion: appcat.vshn.io/v1
kind: PostgreSQL
metadata:
  name: my-instance
spec:
  parameters:
    name: my-instance
    version: '14'
    pgSettings: {}

apiVersion: exoscale.appcat.vshn.io/v1
kind: PostgreSQLConfig
metadata:
  name: my-instance
spec:
  parameters:
    instanceRef: my-instance
    plan: hobbyist-2
    type: pg
    maintenance:
      day: 2
      hour: 22
      minute: 0
    backup:
      hour: 20
      minute: 0
----

Advantages::

* Clear separation of base and specific configuration parameters

Disadvantages::

* "same look" will only apply to the base XRDs
* Pretty complicated to use for the end-user

=== Proposal 4: Same XRD and Parse the Input

We use the same XRD for every provider.

> Most fields are simply strings, and they will then be parsed in the composition.

Reviewer 2: I don't understand this proposal. How is it different from Proposal 1?

Author: The difference is, that in proposal 1 there would be some provider specific fields. Here, we don't have any specific fields at all.

Reviewer 2: Does that mean the user has to provide the correct string, resp. yaml/json?

Author: Yes, exactly.

I've added the source for this idea to the text as well.

* No conditionals in compositions
* Only a limited set of transformations
* Can't create random values

[NOTE]
This is not a complete list of limitations.

== Requirements

* APIs for a set of similar services feel similar
* Only minimal configuration is needed for a default case
* Unique features of a service provider can be leveraged

== Proposals

=== Proposal 1: Use same XRD with provider specific fields

There's exactly one XRD for a set of services.
This XRD contains a minimal set of common parameters, as well as a field for provider specific configurations.

.Example Claim
[source, yaml]
----
apiVersion: appcat.vshn.io/v1
kind: PostgreSQL
metadata:
  name: my-instance
spec:
  parameters:
    name: my-instance
    majorVersion: '14'
    pgSettings: {}
    exoscale:
      plan: hobbyist-2
      type: pg
      maintenance:
        day: 2
        hour: 22
        minute: 0
      backup:
        hour: 20
        minute: 0
    vshn:
      minorVersion: '4'
----

Advantages::

* We can use the same XRD
* Adding new providers to the list is easy

Disadvantages::

* We don't really have a "common look"

> * The customer needs to know what the underlying service provider is, even for simple instances

Reviewer 2: What's the purpose of mentioning "simple instance"? single-replica? how is that different to a "complex instance"?
```suggestion
* The user needs to know some specifics of the underlying service provider.
```

This will also make it clear at one glance that the given provider doesn't support this feature.

.Example Claim
[source, yaml]
----
# Exoscale example
apiVersion: exoscale.appcat.vshn.io/v1
kind: PostgreSQL
metadata:
  name: my-instance
spec:
  parameters:
    name: my-instance
    service:
      version: '14'
      type: pg
      settings: {}
    size:
      plan: hobbyist-2
    maintenance:
      day: 2
      hour: 22
      minute: 0
    backup:
      hour: 20
      minute: 0

# Converged example
apiVersion: vshn.appcat.vshn.io/v1
kind: PostgreSQL
metadata:
  name: my-instance
spec:
  parameters:
    name: my-instance
    service:
      version: '14.4'
      settings: {}
      pgbouncer:
        settings: {}
    size:
      cpu: 2
      memory: 8Gi
      disk: 200Gi
      encrypted: true
    maintenance:
      day: 2
      hour: 22
      minute: 0
    backup:
      hour: 20
      minute: 0
----

Advantages::

* Same look between all providers, deploying PostgreSQL to any backend feels familiar
* Specifics can be handled in a sensible way

Disadvantages::

* The user could get confused by objects with the same kind

> * The customer needs to know what the underlying service provider is, even for simple instances

Reviewer 2: same here
